Họ tên: Nguyễn Minh Vũ

MSSV: 1712920

# BT03: Các loại bộ nhớ trong CUDA

Bạn có thể tùy ý thêm/xóa cell. Đừng xóa mấy cell có chữ của mình là được.

In [1]:
# Mount Google Drive to this notebook
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


## Biên dịch mã nguồn

In [2]:
# cd to the source folder
%cd /content/drive/MyDrive/Study/CUDA/BT03
!nvcc bt03.cu -o bt03

/content/drive/MyDrive/Study/CUDA/BT03


## Thử nghiệm với những kích thước block khác nhau

In [3]:
!./bt03 in.pnm out.pnm 32 32

**********GPU info**********
Name: Tesla T4
Compute capability: 7.5
Num SMs: 40
Max num threads per SM: 1024
Max num warps per SM: 32
GMEM: 15812263936 bytes
CMEM: 65536 bytes
L2 cache: 4194304 bytes
SMEM / one SM: 65536 bytes
****************************

Image size (width x height): 512 x 512

Kernel 1, block size 32x32, grid size 16x16
Kernel time: 0.361984 ms
Error: 0.000703

Kernel 2, block size 32x32, grid size 16x16
Kernel time: 0.345312 ms
Error: 0.000703

Kernel 3, block size 32x32, grid size 16x16
Kernel time: 0.264608 ms
Error: 0.000703


In [4]:
!./bt03 in.pnm out.pnm 32 16

**********GPU info**********
Name: Tesla T4
Compute capability: 7.5
Num SMs: 40
Max num threads per SM: 1024
Max num warps per SM: 32
GMEM: 15812263936 bytes
CMEM: 65536 bytes
L2 cache: 4194304 bytes
SMEM / one SM: 65536 bytes
****************************

Image size (width x height): 512 x 512

Kernel 1, block size 32x16, grid size 16x32
Kernel time: 0.319456 ms
Error: 0.000703

Kernel 2, block size 32x16, grid size 16x32
Kernel time: 0.312960 ms
Error: 0.000703

Kernel 3, block size 32x16, grid size 16x32
Kernel time: 0.242240 ms
Error: 0.000703


In [5]:
!./bt03 in.pnm out.pnm 16 32

**********GPU info**********
Name: Tesla T4
Compute capability: 7.5
Num SMs: 40
Max num threads per SM: 1024
Max num warps per SM: 32
GMEM: 15812263936 bytes
CMEM: 65536 bytes
L2 cache: 4194304 bytes
SMEM / one SM: 65536 bytes
****************************

Image size (width x height): 512 x 512

Kernel 1, block size 16x32, grid size 32x16
Kernel time: 0.473312 ms
Error: 0.000703

Kernel 2, block size 16x32, grid size 32x16
Kernel time: 0.317408 ms
Error: 0.000703

Kernel 3, block size 16x32, grid size 32x16
Kernel time: 0.247168 ms
Error: 0.000703


In [6]:
!./bt03 in.pnm out.pnm 16 32

**********GPU info**********
Name: Tesla T4
Compute capability: 7.5
Num SMs: 40
Max num threads per SM: 1024
Max num warps per SM: 32
GMEM: 15812263936 bytes
CMEM: 65536 bytes
L2 cache: 4194304 bytes
SMEM / one SM: 65536 bytes
****************************

Image size (width x height): 512 x 512

Kernel 1, block size 16x32, grid size 32x16
Kernel time: 0.468992 ms
Error: 0.000703

Kernel 2, block size 16x32, grid size 32x16
Kernel time: 0.319072 ms
Error: 0.000703

Kernel 3, block size 16x32, grid size 32x16
Kernel time: 0.247776 ms
Error: 0.000703


In [7]:
!./bt03 in.pnm out.pnm 16 16

**********GPU info**********
Name: Tesla T4
Compute capability: 7.5
Num SMs: 40
Max num threads per SM: 1024
Max num warps per SM: 32
GMEM: 15812263936 bytes
CMEM: 65536 bytes
L2 cache: 4194304 bytes
SMEM / one SM: 65536 bytes
****************************

Image size (width x height): 512 x 512

Kernel 1, block size 16x16, grid size 32x32
Kernel time: 0.465088 ms
Error: 0.000703

Kernel 2, block size 16x16, grid size 32x32
Kernel time: 0.315392 ms
Error: 0.000703

Kernel 3, block size 16x16, grid size 32x32
Kernel time: 0.242624 ms
Error: 0.000703


### Thử nghiệm thêm trên một ảnh nhỏ hơn

In [8]:
!./bt03 snail.pnm snail_out.pnm 32 32

**********GPU info**********
Name: Tesla T4
Compute capability: 7.5
Num SMs: 40
Max num threads per SM: 1024
Max num warps per SM: 32
GMEM: 15812263936 bytes
CMEM: 65536 bytes
L2 cache: 4194304 bytes
SMEM / one SM: 65536 bytes
****************************

Image size (width x height): 256 x 256

Kernel 1, block size 32x32, grid size 8x8
Kernel time: 0.213152 ms
Error: 0.000275

Kernel 2, block size 32x32, grid size 8x8
Kernel time: 0.114688 ms
Error: 0.000275

Kernel 3, block size 32x32, grid size 8x8
Kernel time: 0.108192 ms
Error: 0.000275


In [9]:
!./bt03 snail.pnm snail_out.pnm 32 16

**********GPU info**********
Name: Tesla T4
Compute capability: 7.5
Num SMs: 40
Max num threads per SM: 1024
Max num warps per SM: 32
GMEM: 15812263936 bytes
CMEM: 65536 bytes
L2 cache: 4194304 bytes
SMEM / one SM: 65536 bytes
****************************

Image size (width x height): 256 x 256

Kernel 1, block size 32x16, grid size 8x16
Kernel time: 0.117632 ms
Error: 0.000275

Kernel 2, block size 32x16, grid size 8x16
Kernel time: 0.112672 ms
Error: 0.000275

Kernel 3, block size 32x16, grid size 8x16
Kernel time: 0.091264 ms
Error: 0.000275


In [10]:
!./bt03 snail.pnm snail_out.pnm 16 16

**********GPU info**********
Name: Tesla T4
Compute capability: 7.5
Num SMs: 40
Max num threads per SM: 1024
Max num warps per SM: 32
GMEM: 15812263936 bytes
CMEM: 65536 bytes
L2 cache: 4194304 bytes
SMEM / one SM: 65536 bytes
****************************

Image size (width x height): 256 x 256

Kernel 1, block size 16x16, grid size 16x16
Kernel time: 0.147552 ms
Error: 0.000275

Kernel 2, block size 16x16, grid size 16x16
Kernel time: 0.102432 ms
Error: 0.000275

Kernel 3, block size 16x16, grid size 16x16
Kernel time: 0.083328 ms
Error: 0.000275


## Nhận xét và giải thích kết quả chạy

- Dùng SMEM chạy nhanh hơn so với khi dùng GMEM: \\
    Vì toàn bộ dữ liệu của một block, kể cả phần bù, được sao chép vào SMEM trước khi thực hiện tích chập, sau đó các thread trong cùng block sẽ dùng chung dữ liệu này. Mà tốc độ đọc/ghi với SMEM nhanh hơn GMEM. Bên cạnh đó, `SMEM / one SM` tối đa là `65536 bytes`, mà mỗi block chỉ dùng tối đa `4800 bytes SMEM` (`(32 + 8) x (32 + 8) x 3`, khi `blockSize = 32 x 32`, mỗi phần tử `uchar3` có kích thước `24 bits = 3 bytes`) nên không ảnh hưởng đến `Occupancy`.

- Dùng CMEM + SMEM chạy nhanh hơn so với khi chỉ dùng SMEM: \\
    Khi khai báo bộ lọc ở tầm vực toàn cục và lưu nó vào CMEM thì các thread chỉ cần đọc dữ liệu này từ `Const Cache` của SM. Mà tốc độ đọc từ `Const Cache` nhanh hơn so với `DRAM`.

## Giải thích thêm về mã nguồn

Tuy mã nguồn đã viết không `hard-code` giá trị nào nhưng nó được tối ưu cho ảnh vuông. Do đó, nếu dùng mã nguồn này với ảnh hình chữ nhật thì sẽ chỉ đúng với một vài `blockSize` (xem các cell code bên dưới).

Thuật toán ghi dữ liệu lên SMEM như sau: mỗi thread ngoài việc sao chép dữ liệu của chính nó vào SMEM thì còn chịu trách nhiệm kiểm tra và sao chép dữ liệu cho 4 điểm khác theo đường chéo trong bán kính `filterWidth / 2` (trên bên trái, trên bên phải, dưới bên trái, dưới bên phải). Mỗi thread cần trải qua 4 lần kiểm tra điều kiện thì mới hoàn thành quá trình sao chép dữ liệu.

Trước đó, em đã làm theo cách chỉ các thread ở biên của block mới đảm nhận việc sao chép dữ liệu, do đó phải sử dụng đến vòng lặp. Mà bởi chi phí của vòng lặp lớn nên dùng SMEM lại cho kết quả xấu hơn chỉ dùng GMEM.

In [11]:
!./bt03 cat.pnm cat_out.pnm 32 32

**********GPU info**********
Name: Tesla T4
Compute capability: 7.5
Num SMs: 40
Max num threads per SM: 1024
Max num warps per SM: 32
GMEM: 15812263936 bytes
CMEM: 65536 bytes
L2 cache: 4194304 bytes
SMEM / one SM: 65536 bytes
****************************

Image size (width x height): 240 x 320

Kernel 1, block size 32x32, grid size 8x10
Kernel time: 0.118784 ms
Error: 0.000599

Kernel 2, block size 32x32, grid size 8x10
Kernel time: 0.114720 ms
Error: 0.081949

Kernel 3, block size 32x32, grid size 8x10
Kernel time: 0.100352 ms
Error: 0.443255


In [12]:
!./bt03 cat.pnm cat_out.pnm 32 16

**********GPU info**********
Name: Tesla T4
Compute capability: 7.5
Num SMs: 40
Max num threads per SM: 1024
Max num warps per SM: 32
GMEM: 15812263936 bytes
CMEM: 65536 bytes
L2 cache: 4194304 bytes
SMEM / one SM: 65536 bytes
****************************

Image size (width x height): 240 x 320

Kernel 1, block size 32x16, grid size 8x20
Kernel time: 0.130048 ms
Error: 0.000599

Kernel 2, block size 32x16, grid size 8x20
Kernel time: 0.112832 ms
Error: 0.000599

Kernel 3, block size 32x16, grid size 8x20
Kernel time: 0.089920 ms
Error: 0.000599


In [13]:
!./bt03 cat.pnm cat_out.pnm 16 32

**********GPU info**********
Name: Tesla T4
Compute capability: 7.5
Num SMs: 40
Max num threads per SM: 1024
Max num warps per SM: 32
GMEM: 15812263936 bytes
CMEM: 65536 bytes
L2 cache: 4194304 bytes
SMEM / one SM: 65536 bytes
****************************

Image size (width x height): 240 x 320

Kernel 1, block size 16x32, grid size 15x10
Kernel time: 0.149152 ms
Error: 0.000599

Kernel 2, block size 16x32, grid size 15x10
Kernel time: 0.113280 ms
Error: 0.000599

Kernel 3, block size 16x32, grid size 15x10
Kernel time: 0.092128 ms
Error: 0.000599
